### PR TITLE
A solution to minimise the nlpost

### PR DESCRIPTION
**Pull request by arm61**
A possible (pragmatic) solution to https://github.com/refnx/refnx/issues/318

### CODE REVIEW

**Comment by arm61** [comment]
Let me know thoughts on this style of solution and I will get some unit tests sorted.

File: refnx/analysis/curvefitter.py [hunk 1]
```diff
@@ -535,7 +535,7 @@ def _callback_wrapper(state, h=None):
         # sets parameter value and stderr
         return process_chain(self.objective, self.chain)
 
-    def fit(self, method='L-BFGS-B', **kws):
+    def fit(self, method='L-BFGS-B', prior=False, **kws):
```
Comment (andyfaff) [inline refnx/analysis/curvefitter.py]: I'd prefer to use a *slightly* different approach. I'm not much good with naming, but how about we use a `target='nll'` keyword. The options for the `target` keyword would be `{'nll', 'nlpost'}`.
Default would be to do the maximum log-likelihood, but the other option would be maximise the log-posterior (often referred to as MAP).
I think that `nlpost` will fail if gradient based minimisers stray into an area for which the prior is non-finite (`-np.inf`), otherwise they should be fine. Differential evolution should be able to cope with non-finite priors so long as they're not ubiquitous.

I think the biggest challenge is the use of a `PDF` object to specify the prior/bounds for a given parameter. Let's say the `PDF` is Gaussian. It's easy to calculate the probabilities for that. The issue comes when an optimisation method (e.g. differential evolution) mandates lower and upper bounds for the parameter. The fitter [looks to each parameter](https://github.com/refnx/refnx/blob/master/refnx/analysis/curvefitter.py#L846) to see if its bounds attribute is an `Interval`. If it's not an `Interval`, which is won't be if you use a `PDF`, then the fitting process won't work because there will be no lower and upper bounds. I'm not sure how to get around this challenge, it'll require some thought.

TLDR: Not sure if it's possible to use anything other than a non-uniform distribution with a method that works on box (lower, upper) bounds.

File: refnx/analysis/curvefitter.py [hunk 2]
```diff
@@ -555,6 +555,11 @@ def fit(self, method='L-BFGS-B', **kws):
 
             You can also choose many of the minimizers from
             ``scipy.optimize.minimize``.
+        prior : bool
```
Comment (andyfaff) [inline refnx/analysis/curvefitter.py]: `target : {'nll', 'nlpost'}, optional`

File: refnx/analysis/curvefitter.py [hunk 3]
```diff
@@ -555,6 +555,11 @@ def fit(self, method='L-BFGS-B', **kws):
...
+            Have prior probabilities been defined that should be considered?
```
Comment (andyfaff) [inline refnx/analysis/curvefitter.py]: docstring amended accordingly.

File: refnx/analysis/curvefitter.py (hunk 1 above)
Comment (andyfaff) [reply refnx/analysis/curvefitter.py]: It's only going to work on the scalar minimisers, not least-squares.

File: refnx/analysis/curvefitter.py [hunk 4]
```diff
@@ -555,6 +555,11 @@ def fit(self, method='L-BFGS-B', **kws):
...
+            This is only supported for the differential_evolution optimisation
```
Comment (andyfaff) [inline refnx/analysis/curvefitter.py]: Should work for any of the scalar minimisers, but not least-squares.

File: refnx/analysis/curvefitter.py [hunk 5]
```diff
@@ -604,7 +609,16 @@ def fit(self, method='L-BFGS-B', **kws):
         # bounds
         elif method in ['differential_evolution', 'dual_annealing', 'shgo']:
             mini = getattr(sciopt, method)
-            res = mini(self.objective.nll, **_min_kws)
+            if prior:
```
Comment (andyfaff) [inline refnx/analysis/curvefitter.py]: These additions aren't needed.

File: refnx/analysis/curvefitter.py [hunk 6]
```diff
@@ -604,7 +609,16 @@ def fit(self, method='L-BFGS-B', **kws):
...
+                if method == 'differential_evolution':
+                    res = mini(self.objective.nlpost, **_min_kws)
+                else:
+                    raise ValueError('The use of a prior probability is '
+                                     'currently only supported for the '
+                                     'differential_evolution optimisation '
+                                     'method.')
+            else:
+                res = mini(self.objective.nll, **_min_kws)
```
Comment (andyfaff) [inline refnx/analysis/curvefitter.py]: All you need to do is:

```
cost = self.objective.nll
if target == 'nlpost':
    cost = self.objective.nlpost

res = mini(cost, **_min_kws)
```

**Comment by andyfaff** [comment]
There will be challenges for non-uniform priors for the scalar minimisers which use box (lower, upper) bounds. However, I'd make the changes requested here and we can cross that bridge when we come to it.

**Comment by andyfaff** [comment]
I have an idea of how to approximate bounds for a MAP. `scipy.stats` distributions have various methods, one of which is the [`ppf` function](https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.rv_continuous.ppf.html).

We could estimate that the lower and upper bound is at e.g. 0.01 and 0.99 of the cumulative distribution function (CDF). For example, the following code works out the x-values for 16 % and 84% of the way through the normal distribution. These should be -1 and 1 respectively because those values lie 1 s.d. away from the mean.

```
>>> from scipy.stats import norm
>>> print(norm.ppf([0.16, 0.84]))
[-0.99445788  0.99445788]
```

Thus, the code in the `fit` method could be:

- see if it's an nlpost or nll fit.
- Go through each of the parameters and ensure they all have `Bound`s. If the `Bound`s are `Interval` it's easy to setup lower and upper limit. If the `Bound` is a `PDF` then estimate a lower and upper limit by trying to use `PDF.ppf` with e.g. 0.01 and 0.99 of the CDF.

This will fail to find the maximum probability if the value lies in the tails though.

**Comment by arm61** [comment]
This is a nice general solution!

Sorry for the delay in sorting it myself (I had my viva on Thursday and have been ignoring work since then) but I think that your solution might be more general than the method I had pieced together.
